Add Static code analysis tool to ESH build

svilenvul opened the pull request:
Closes #2437 

Before merging:
 - ~~CQ should be approved~~;
 - ~~the version of the tool has to be updated with the latest that is released~~; 
 - ~~the high priority issues should be fixed~~;
 - ~~https://github.com/eclipse/smarthome/issues/3963 is resolved. See https://github.com/openhab/static-code-analysis/issues/136~~

htreu commented on tools/static-code-analysis/checkstyle/suppressions.xml:
```diff
@@ -16,4 +16,6 @@
     <!-- These bundles are generated trough XText -->
     <suppress files=".+org.eclipse.smarthome.model.+" checks="RequireBundleCheck|ExportInternalPackageCheck|ManifestPackageVersionCheck|ImportExportedPackagesCheck"/>
+    <!-- Some source files have different headers -->
+    <suppress files=".+org.eclipse.smarthome.automation.+" checks="HeaderCheck"/>
```
Do we really want to suppress checks or rather update the automation bundles to adhere to our guidelines?

kaikreuzer replied:
Yes, because we currently have a different arrangement for the license header of the automation component.

htreu replied:
okay, tnx.

maggu2810 commented on extensions/binding/org.eclipse.smarthome.binding.hue.test/src/test/groovy/org/eclipse/smarthome/binding/hue/test/HueBridgeHandlerOSGiTest.groovy:
```diff
@@ -218,7 +218,7 @@ class HueBridgeHandlerOSGiTest extends AbstractHueOSGiTest {
 
     @Test
-    void 'assert that a status configuration message for missing bridge IP is properly returned (IP is an empty string)'() {
+    void 'assert that a status configuration message for missing bridge IP is properly returned - IP is an empty string'() {
```
Can you create a separate PR for that change?
I expect that it will be okay to change the long groovy names.

maggu2810 commented on extensions/binding/org.eclipse.smarthome.binding.hue.test/src/test/groovy/org/eclipse/smarthome/binding/hue/test/HueBridgeHandlerOSGiTest.groovy:
```diff
@@ -197,7 +197,7 @@ class HueBridgeHandlerOSGiTest extends AbstractHueOSGiTest {
 
     @Test
-    void 'assert that a status configuration message for missing bridge IP is properly returned (IP is null)'() {
+    void 'assert that a status configuration message for missing bridge IP is properly returned - IP is null'() {
```
Can you create a separate PR for that change?
I expect that it will be okay to change the long groovy names.

svilenvul replied:
Yes, of course.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java:
```diff
@@ -78,7 +78,7 @@ public static void assertValidItemName(String itemName) throws IllegalArgumentEx
     public static State convertToAcceptedState(State state, Item item) {
         if (state == null) {
-            LOGGER.error("A conversion of null was requested", new NullPointerException("state should not be null"));
+            LOGGER.error("A conversion of null was requested", new NullPointerException("state should not be null")); // NOPMD
```
What is the exact PMD warning that is suppressed here?

svilenvul replied:
See https://github.com/eclipse/smarthome/pull/3888

kaikreuzer commented:
Created CQ 13980 for the tool dependency.

kaikreuzer commented:
Woah, that was quick: The CQ is already approved!

svilenvul commented:
The latest build fails with the hue binding. After the update of the XML things schema, the build should be green :-).

svilenvul commented:
@kaikreuzer 

Friday the only remaining errors were caused from the .xsd file that is not up to date - #3963.

Should I ignore these errors or wait for the update of the schema file ?

kaikreuzer commented:
Wait for the schema file - it should be up to date now! I retrigger Travis...

svilenvul commented:
Retrigger Travis.

svilenvul commented:
The build failed after 49 mins :
```
...
Still running (49 of 60): ./.build.sh
Still running (50 of 60): ./.build.sh

The job exceeded the maximum time limit for jobs, and has been terminated.
```
@kaikreuzer I guess the build is too slow with the static-analysis-tool added.

sjka commented:
I restarted it once again, just to make sure it was not caused by a slow repository, but the result was the same. That's indeed not so good news...
The global timeout for travis builds indeed is 50 minutes (see https://docs.travis-ci.com/user/customizing-the-build#Build-Timeouts), so there is nothing much we can do about it. 
The build until now took something just above 20 minutes. Is it at least plausible that the code analysis adds _that_ much to it? Sounds like a lot, so I'm wondering whether something went seriously wrong or not...

svilenvul commented:
When we talk about speed/performance, FindBugs is always the main suspected :-) .

```
mvn static-code-analysis:findbugs
....
[INFO] Total time: 18:46 min
[INFO] Finished at: 2017-08-15T11:45:17+03:00
[INFO] Final Memory: 238M/1127M
```
Checkstyle checks are taking 06:32 min and PMD 02:41 min.

I guess we can speed it up a bit if we tweak the configuration of FindBugs a little bit, but I am not sure if it will be enough.

I would appreciate help here.

svilenvul commented:
I have tried several scenarios to reduce the time that is needed from FindBugs without much success:
- applied some improvements to the configuration - reduced the time with 2 minutes;
- reduced the number of detectors from 60 to 3 (the one that log the most errors so far) :-) and FindBugs still needs 12 minutes to check all bundles (that is only 6 minutes less build time).

I have carefully looked at the maven log (with debug activated for FindBugs) and I see no unexpected behavior. I will not attach here the logs here as they are 100 MB.

When the timeout is fixed to 50 minutes, I would propose to disable the `findbugs` goal at all. All the custom checks that we have are implemented for Checkstyle. The FindBugs detectors are not following the coding guidelines (with the exception of the logger detectors), they are extra checks in this case. We will keep the main functionality of the tool without FindBugs.

With FindBugs disabled my local build has finished for 23 minutes: 
```
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 23:28 min
[INFO] Finished at: 2017-08-15T17:15:02+03:00
[INFO] Final Memory: 626M/1894M
[INFO] ------------------------------------------------------------------------
```

Please let me know what do you think about disabling the FindBugs checks for the ESH repository.

maggu2810 commented:
Instead of disabling checks we should consider to start the split of core and extensions into different repositories. This will reduce the build time for the respective parts and perhaps allow to run all tests.

kaikreuzer commented:
But it also means more overhead in maintaining the builds, so I am not too keen on that.
Just to prove the point: When was the last time that we have build a Designer, since it has been moved to its own repo...?

J-N-K commented:
Maybe that has something to do with the general state of the designer (i.e. the fact that the current version is quite useless). However, this should be decided my those that have to maintain the repos.

kaikreuzer commented:
@J-N-K Its days are numbered anyhow, see https://dev.eclipse.org/mhonarc/lists/smarthome-dev/msg00146.html.

J-N-K commented:
@kaikreuzer, that is what I wanted to say. The designer is in bad shape and not much development is being done there (because better alternatives are on their way, which is good). Therefore I think the designer is not a good comparison. IMHO splitting extensions and core makes more modular solutions possible (e.g. core+specialized extension, leaving out the "standard" extensions).

maggu2810 commented:
> But it also means more overhead in maintaining the builds, so I am not too keen on that.

Okay, then I missed some news. AFAIK you prefer a split of core and extensions.

> Just to prove the point: When was the last time that we have build a Designer, since it has been moved to its own repo...?

IMHO it is more about that the designer is not really maintained and not because it has been moved to another repo.

maggu2810 commented:
@svilenvul I stumbled over a file using non-common modifier order. Shouldn't we force the usage of the common order?

svilenvul commented:
We have a check for modifier order - https://github.com/openhab/static-code-analysis/commit/fca697a729df9dcd253ea9983ef07954c933d98b.

svilenvul commented:
I can suggest more more "workarounds":
- additional Jenkins profile (fast profile) with FindBugs disabled ?

svilenvul commented:
@kaikreuzer, I guess it is up to you. 

> Please let me know what do you think about disabling the FindBugs checks for the ESH repository.

What we are going to do in this situation ?

kaikreuzer commented:
> Okay, then I missed some news. AFAIK you prefer a split of core and extensions -- some time ago.

No, you are only always underestimating my pragmatism. While I agree that it would be nice to have it in a separate repo and also to version it separately from the core, I have learned from having openHAB split into several repos what overhead this means in terms of build & release management. After all, we duplicate the number of build jobs, the repos to update for new tools like the static code analysis or other upgrades, the efforts on doing releases and the number of issue trackers (while people today already never know where to post something).
Seeing that we didn't manage to do an official release for a long time let's me doubt that we are in a position to think about separate versioning of core and extensions (which would imho be the main reason for a repo split).

> Please let me know what do you think about disabling the FindBugs checks for the ESH repository.

I don't think that genereally disabling valuable checks is an option. We should rather think about how to reduce the situations, where we execute them. I would claim that the build time of the nightly job on our hudson is not an issue, so I would prefer to have it executing the full test/check suite.

Wrt Travis, I am not sure what our options are. If you could add a parameter which deactivates the FindBugs checks and we set that in the Travis configuration only, it might be an acceptable compromise.

svilenvul commented:
Great, I have deactivated the FindBugs checks only in Travis.

This has caused [some errors](https://github.com/openhab/static-code-analysis/commit/6a63fc3e60163734a62def6385988bbd0ae93e01#commitcomment-23718985) in the report generation goal, but they are already fixed with this [commit](https://github.com/openhab/static-code-analysis/commit/6a63fc3e60163734a62def6385988bbd0ae93e01).

kaikreuzer commented:
> so maybe it is better to release new version first and then add it to the ESH build.

New version 0.3.0 has been released - do you want to update the PR accordingly? I think we would be good to merge then!

svilenvul commented:
@kaikreuzer, I have rebased and updated the PR.

Do you think we should add a few sentences for the tool in the ESH documentation like we have done this in the openHAB docs - http://docs.openhab.org/developers/development/bindings.html#static-code-analysis ? Where would be the best place for that - maybe in the Coding Guidelines ?

kaikreuzer commented:
Yes, I guess the coding guidelines are the best place for it.

svilenvul commented:
@kaikreuzer, I am ready with the updates on this PR.

maggu2810 commented on docs/documentation/development/guidelines.md:
```diff
@@ -68,3 +68,11 @@ Note that this list also serves as a checklist for code reviews on pull requests
+
+The Eclipse SmartHome Maven build includes [tooling for static code analysis](https://github.com/openhab/static-code-analysis) that will validate your code against the Coding Guidelines and some additional best practices. Information about the checks can be found [here](https://github.com/openhab/static-code-analysis/blob/master/docs/included-checks.md).
+
+The tool will generate an idividual report for each binding that you can find in `../your_binding_directory/target/code-analysis/report.html` file and a report for the whole build that contains links to the individual reports in the `../smarthome/target/summary_report.html`. The tool categorizes the found issues by priority: 1(error),2(warning) or 3(info). If any error is found within your code t ...
```
It is not only about bindings but every bundle that is build (for each binding => for each bundle). And to be fair it would be not only be used for bundles, but for every artifact in the reactor, so "for each artifact" would be correct.

svilenvul replied:
At the moment it will be used only on bundles - there are two profiles in this PR that activate the tool for bundles and bindings (the one is obsolete now, I will remove it). Initially the idea was to have separate rulesets included in each profile, but this changed over time. 

Activating the checks for each artifact is not tested well enough, some of the OSGi specific checks might have problems with that.

maggu2810 commented on docs/documentation/development/guidelines.md:
```diff
@@ -68,3 +68,11 @@ Note that this list also serves as a checklist for code reviews on pull requests
+
+The Eclipse SmartHome Maven build includes [tooling for static code analysis](https://github.com/openhab/static-code-analysis) that will validate your code against the Coding Guidelines and some additional best practices. Information about the checks can be found [here](https://github.com/openhab/static-code-analysis/blob/master/docs/included-checks.md).
+
+The tool will generate an idividual report for each binding that you can find in `../your_binding_directory/target/code-analysis/report.html` file and a report for the whole build that contains links to the individual reports in the `../smarthome/target/summary_report.html`. The tool categorizes the found issues by priority: 1(error),2(warning) or 3(info). If any error is found within your code t ...
```
Why do you add a `../` in front of the path.
Normally you call `mvn ...` in the root of the reactor and IMHO the relative paths should be relative to the root.

svilenvul replied:
Thanks, I was wondering how to be more precise here, but without much success. I will change it to relative to the root of the reactor.

maggu2810 commented on pom.xml:
```diff
@@ -469,6 +470,120 @@
+    </pluginRepository>
+    <pluginRepository>
+      <id>buchen-maven-repo</id>
+      <url>http://buchen.github.io/maven-repo</url>
```
Hasen't this one been removed previously by intention?
Why do you add it again?

maggu2810 commented on pom.xml:
```diff
@@ -469,6 +470,120 @@
+      <id>eclipse</id>
+      <name>Eclipse Snapshot Repository</name>
+      <layout>default</layout>
+      <url>https://repo.eclipse.org/content/repositories/snapshots/</url>
```
This has been removed some days ago, why do you need the Eclipse Snapshot Repository?

svilenvul replied:
As I see I only need both JFrog repositories (one for the plugin and one for its dependencies) , I will remove the others.

svilenvul commented:
I hope I have answered all questions. Do you have any other concerns ?

kaikreuzer commented on pom.xml:
```diff
@@ -469,6 +470,81 @@
 
+  <pluginRepositories>
+    <pluginRepository>
+      <id>openhab-artifactory-release</id>
```
Actually, the official repo for the releases is "jcenter" at https://jcenter.bintray.com/.

It probably makes sense that I also request to have it published to Maven central, then we would not at all need a specific repo here.

kaikreuzer commented on pom.xml:
```diff
@@ -469,6 +470,81 @@
+  </pluginRepositories>
+
+  <repositories>
+    <repository>
```
Isn't the pluginRepository enough? Why do we need it as a dependency repository as well?

svilenvul replied:
Good question, I haven't taught about that much until you asked. 
The tool(as a Maven *plugin*) internally starts the Maven plugins for Checkstyle, PMD and FindBugs and adds itself as a *dependency* to these plugins, so the included custom checks that are contained there are available in the classpath.

kaikreuzer replied:
Did you test if it also works without it?

svilenvul replied:
Yes, I tested it, it is not working without this repository.

kaikreuzer commented on pom.xml:
```diff
@@ -469,6 +470,81 @@
 
+  <pluginRepositories>
+    <pluginRepository>
+      <id>JCenter Repository</id>
```
id should imho rather be "jcenter"

kaikreuzer commented on pom.xml:
```diff
@@ -469,6 +470,81 @@
+
+  <repositories>
+    <repository>
+      <id>openhab-artifactory-release</id>
```
id should be "jcenter"

kaikreuzer commented:
Thanks!
